source/common/tracing/zipkin/zipkin_core_types.h

```diff
@@ -0,0 +1,536 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/optional.h"
+#include "envoy/common/pure.h"
+#include "envoy/network/address.h"
+
+#include "common/common/hex.h"
+#include "common/tracing/zipkin/tracer_interface.h"
+#include "common/tracing/zipkin/util.h"
+
+namespace Zipkin {
+
+/**
+ * Base class to be inherited by all classes that represent Zipkin-related concepts, namely:
+ * endpoint, annotation, binary annotation, and span.
+ */
+class ZipkinBase {
+public:
+  /**
+   * Destructor.
+   */
+  virtual ~ZipkinBase() {}
+
+  /**
+   * All classes defining Zipkin abstractions need to implement this method to convert
+   * the corresponding abstraction to a Zipkin-compliant JSON.
+   */
+  virtual const std::string& toJson() PURE;
+
+protected:
+  std::string json_string_;
```

Comment:
It's pretty odd to have this be a member variable and basically use it as a temporary. I would get rid of this, and then have all the functions return a temporary. The compiler will optimize it.